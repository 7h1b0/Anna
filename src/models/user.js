const mongoose = require('mongoose');
const Ajv = require('ajv');
const userSchema = require('../schemas/user');

const Schema = mongoose.Schema;

const user = new Schema({
  username: { type: String, required: true, unique: true },
  password: { type: String, required: true },
  token: { type: String, required: true },
});

user.statics.validate = function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(userSchema, data);
};

module.exports = mongoose.model('User', user);
