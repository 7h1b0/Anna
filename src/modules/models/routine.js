import Ajv from 'ajv';
import knex from '../../knexClient';
import routineSchema from '../schemas/routine';
import { returnFirst } from '../dbUtil';

export const TABLE = 'routines';
export const COLUMNS = [
  { routineId: 'routine_id' },
  'name',
  'schedule',
  { sceneId: 'scene_id' },
  { runAtBankHoliday: 'run_at_bank_holiday' },
  'enabled',
  { createdAt: 'created_at' },
  { updatedAt: 'updated_at' },
  { failedAt: 'failed_at' },
  { lastRunAt: 'last_run_at' },
  { nextRunAt: 'next_run_at' },
  { createdBy: 'created_by' },
];

export function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(routineSchema, data);
}

export function findAll() {
  return knex(TABLE).select(COLUMNS);
}

export function findById(routineId) {
  return returnFirst(
    knex(TABLE)
      .select(COLUMNS)
      .where('routine_id', routineId),
  );
}

export function save(
  userId,
  name,
  sceneId,
  schedule,
  enabled = true,
  runAtBankHoliday = true,
) {
  return returnFirst(
    knex(TABLE).insert({
      scene_id: sceneId,
      name,
      schedule,
      enabled,
      run_at_bank_holiday: runAtBankHoliday,
      created_by: userId,
    }),
  );
}

export function remove(routineId) {
  return knex(TABLE)
    .where('routine_id', routineId)
    .del();
}

export function findByIdAndUpdate(routineId, payload) {
  return knex(TABLE)
    .update({ ...payload, updated_at: new Date() })
    .where('routine_id', routineId);
}
