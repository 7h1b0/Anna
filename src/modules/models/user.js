import Ajv from 'ajv';
import userSchema from '../schemas/user';
import knex from '../../knexClient';
import { returnFirst } from '../dbUtil';

export const TABLE = 'users';

export function findByToken(token) {
  return returnFirst(
    knex(TABLE)
      .select({ userId: 'user_id' }, 'username')
      .where('token', token),
  );
}

export function findAll() {
  return knex(TABLE).select({ userId: 'user_id' }, 'username');
}

export function findByUsername(username) {
  return returnFirst(
    knex(TABLE)
      .select({ userId: 'user_id' }, 'username', 'password', 'token')
      .where('username', username),
  );
}

export function findByIdAndUpdate(userId, payload) {
  return knex(TABLE)
    .update(payload)
    .where('user_id', userId);
}

export function save({ username, password, token }) {
  return knex(TABLE)
    .insert({ username, password, token })
    .then(([userId]) => {
      return { userId, token, username };
    });
}

export function remove(userId) {
  return knex(TABLE)
    .where('user_id', '=', userId)
    .del();
}

export function validate(data) {
  const ajv = new Ajv();
  return ajv.validate(userSchema, data);
}
