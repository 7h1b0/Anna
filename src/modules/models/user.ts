import Ajv from 'ajv';
import uuidv4 from 'uuid/v4';
import userSchema from '../schemas/user';
import knex from '../../knexClient';
import { omit } from '../utils';

export const TABLE = 'users';

export interface User {
  userId: string;
  username: string;
  password: string;
  token: string;
}

export function findByToken(token: string) {
  return knex(TABLE)
    .first('userId', 'username')
    .where('token', token);
}

export async function findAll(): Promise<{ userId: string; username: string }> {
  return knex(TABLE).select('userId', 'username');
}

export async function findByUsername(username: string): Promise<User> {
  return knex(TABLE)
    .first('userId', 'username', 'password', 'token')
    .where('username', username);
}

export function findByIdAndUpdate(userId: string, payload) {
  const safePayload = omit(payload, ['userId']);
  return knex(TABLE)
    .update(safePayload)
    .where('userId', userId);
}

export async function save(newUser: Pick<User, Exclude<keyof User, 'userId'>>) {
  const userId = uuidv4();
  await knex(TABLE).insert({ ...newUser, userId });
  return userId;
}

export function remove(userId: string) {
  return knex(TABLE)
    .where('userId', userId)
    .del();
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(userSchema, data);
}
