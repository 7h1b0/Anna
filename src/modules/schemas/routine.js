export default {
  $id: 'routine',
  type: 'object',
  additionalProperties: false,
  properties: {
    routineId: {
      type: 'string',
      format: 'uuid',
    },
    name: {
      type: 'string',
      minLength: 3,
    },
    interval: { type: 'string' },
    sceneId: {
      type: 'string',
      format: 'uuid',
    },
    enabled: { type: 'boolean' },
    runAtBankHoliday: { type: 'boolean' },
    createdAt: { type: 'integer' },
    updatedAt: { type: 'integer' },
    failedAt: { type: 'integer' },
    lastRunAt: { type: 'integer' },
    nextRunAt: { type: 'integer' },
    createdBy: {
      type: 'string',
      format: 'uuid',
    },
  },
  required: ['name', 'sceneId', 'interval'],
};
