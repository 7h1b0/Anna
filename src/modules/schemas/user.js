module.exports = {
  $id: 'user',
  type: 'object',
  additionalProperties: false,
  properties: {
    username: {
      type: 'string',
      minLength: 4,
    },
    password: {
      type: 'string',
      minLength: 4,
    },
  },
  required: ['username', 'password'],
};
