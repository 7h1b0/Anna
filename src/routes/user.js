const User = require('../modules/models/user');
const { cryptoUtil } = require('../modules/');

module.exports = app => {
  app.post('/register', (req, res) => {
    const isValid = User.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }
    const { username, password } = req.body;
    Promise.all([cryptoUtil.random(36), cryptoUtil.hash(password)])
      .then(([token, hashedPassword]) => {
        return User.save({
          username,
          password: hashedPassword,
          token,
        });
      })
      .then(newUser => {
        return res.status(201).json(newUser);
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  });

  app.post('/login', (req, res) => {
    const isValid = User.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }
    const { username, password } = req.body;
    User.findByUsername(username)
      .then(findUser => {
        if (!findUser) {
          res.sendStatus(403);
        } else {
          cryptoUtil
            .verify(password, findUser.password)
            .then(isValid => {
              if (!isValid) {
                throw new Error('Wrong password');
              }
              const copyUser = {
                _id: findUser.user_id,
                username: findUser.username,
                token: findUser.token,
              };

              res.json(copyUser);
            })
            .catch(() => res.sendStatus(403));
        }
      })
      .catch(err => res.status(500).send({ err }));
  });

  app.get('/api/users', (req, res) => {
    User.findAll()
      .then(users => res.send(users))
      .catch(err => res.status(500).send({ err }));
  });

  app
    .route('/api/users/:id_user([0-9]+)')
    .patch((req, res) => {
      const isValid = User.validate(req.body);
      if (!isValid) {
        return res.sendStatus(400);
      }

      const { password } = req.body;
      cryptoUtil
        .hash(password)
        .then(hashedPassword =>
          User.findByIdAndUpdate(req.params.id_user, {
            password: hashedPassword,
          }),
        )
        .then(rowsAffected => {
          if (rowsAffected < 1) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch(err => res.status(500).send({ err }));
    })
    .delete((req, res) => {
      User.delete(req.params.id_user)
        .then(user => {
          if (!user) {
            res.sendStatus(404);
          } else {
            res.sendStatus(204);
          }
        })
        .catch(err => res.status(500).send({ err }));
    });
};
