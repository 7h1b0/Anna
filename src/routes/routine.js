import { Router } from 'express';
import * as Routine from '../modules/models/routine';
import * as RoutineService from '../services/routineService';
import * as logger from '../modules/logger';

const routes = Router();
routes
  .route('/api/routines')
  .get((req, res) => {
    Routine.findAll()
      .then(routines => res.json(routines))
      .catch(err => res.status(500).send({ err }));
  })
  .post((req, res) => {
    const isValid = Routine.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }

    const userId = res.locals.user.userId;
    const {
      name,
      sceneId,
      interval,
      enabled = true,
      runAtBankHoliday = true,
    } = req.body;

    Routine.save(userId, name, sceneId, interval, enabled, runAtBankHoliday)
      .then(routineId => {
        RoutineService.start({
          routineId,
          name,
          sceneId,
          interval,
          enabled,
          runAtBankHoliday,
        });
        res.status(201).json({ routineId });
      })
      .catch(err => res.status(500).send({ err }));
  });

routes
  .route('/api/routines/:routineId([a-fA-F0-9-]{36})')
  .get((req, res) => {
    Routine.findById(req.params.routineId)
      .then(routine => {
        if (!routine) {
          res.sendStatus(404);
        } else {
          res.json(routine);
        }
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  })
  .patch((req, res) => {
    const isValid = Routine.validate(req.body);
    if (!isValid) {
      return res.sendStatus(400);
    }
    const routineId = req.params.routineId;
    Routine.findByIdAndUpdate(routineId, req.body)
      .then(async rowsAffected => {
        if (!!rowsAffected) {
          const { updatedAt, createdAt, ...routine } = await Routine.findById(
            routineId,
          );
          RoutineService.start(routine);
          res.sendStatus(204);
        } else {
          res.sendStatus(404);
        }
      })
      .catch(err => res.status(500).send({ err }));
  })
  .delete((req, res) => {
    Routine.remove(req.params.routineId)
      .then(removedScene => {
        if (removedScene < 1) {
          res.sendStatus(404);
        } else {
          RoutineService.stop(req.params.routineId);
          res.sendStatus(204);
        }
      })
      .catch(err => {
        res.status(500).send({ err });
      });
  });

routes.get('/api/routines/:routineId([a-fA-F0-9-]{36})/action', (req, res) => {
  Routine.findById(req.params.routineId)
    .then(routine => {
      if (routine) {
        return Routine.run(routine);
      }
      throw new Error('Routine not found');
    })
    .then(() => res.end())
    .catch(err => {
      if (err.message === 'Routine not found') {
        return res.sendStatus(404);
      }
      logger.error(err);
      res.status(500).send({ err });
    });
});

export default routes;
