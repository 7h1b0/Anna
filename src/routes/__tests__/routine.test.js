import request from 'supertest';
import createUser from 'createUser';
import knex from '../../knexClient';
import * as Routine from '../../modules/models/routine';
import * as User from '../../modules/models/user';
import app from '../../index.js';
import * as RoutineService from '../../services/routineService';

jest.unmock('cron');

const user = createUser({ userId: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f' });
const initRoutines = [
  {
    routineId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 5am',
    interval: '0 5 * * *',
    enabled: true,
    runAtBankHoliday: true,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    routineId: '1fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '10c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 9am',
    interval: '0 9 * * *',
    enabled: true,
    runAtBankHoliday: false,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
];

describe('Routine API', () => {
  beforeAll(async () => {
    await knex(Routine.TABLE).truncate();
    await knex(User.TABLE).insert(user);
  });

  beforeEach(async () => {
    await knex(Routine.TABLE).insert(initRoutines);
  });

  afterEach(async () => {
    await knex(Routine.TABLE).truncate();
  });

  afterAll(async () => {
    await knex(User.TABLE).truncate();
    await knex.destroy();
  });

  describe('/api/routines', () => {
    describe('GET', () => {
      it('should retun all routines', async () => {
        const response = await request(app)
          .get('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });
    });

    describe('POST', () => {
      it('should retun 401 when user is not authenticated', async () => {
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
      });

      it('should retun 400 when routine is invalid', async () => {
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
          });

        expect(response.status).toBe(400);
      });

      it('should save and start a new routine', async () => {
        const spy = jest.spyOn(RoutineService, 'start');
        const response = await request(app)
          .post('/api/routines')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'new_routine',
            interval: '* * * * * *',
          });

        expect(response.status).toBe(201);
        expect(RoutineService.start.mock.calls).toMatchSnapshot();

        const routines = await knex(Routine.TABLE)
          .first()
          .where('routineId', response.body.routineId);

        expect(routines).toMatchSnapshot({
          routineId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        });

        spy.mockRestore();
      });
    });
  });

  describe('/api/routines/:routineId', () => {
    describe('GET', () => {
      it('should return one routine', async () => {
        const response = await request(app)
          .get(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.body).toMatchSnapshot();
      });

      it('should return 404 if routineId is unknow', async () => {
        const response = await request(app)
          .get('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
        expect(response.body).toEqual({});
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .get(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
        expect(response.body).toEqual({});
      });
    });

    describe('PATCH', () => {
      it('should return 404 if routineId is unknow', async () => {
        const response = await request(app)
          .patch('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
            interval: '* * * * * *',
          });

        expect(response.status).toBe(404);
        expect(response.body).toEqual({});
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake')
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
            interval: '* * * * * *',
          });

        expect(response.status).toBe(401);
        expect(response.body).toEqual({});
      });

      it('should return 400 when routine is invalid', async () => {
        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'room_updated',
          });

        expect(response.status).toBe(400);
      });

      it('should update a routine', async () => {
        const spy = jest.spyOn(RoutineService, 'start');

        const response = await request(app)
          .patch(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token)
          .send({
            sceneId: 'faaed78e-fd1c-4717-b610-65d2fa3d01b2',
            name: 'routine_updated',
            interval: '* * * * * *',
          });

        expect(response.status).toBe(204);
        expect(RoutineService.start.mock.calls).toMatchSnapshot();

        const routine = await knex(Routine.TABLE)
          .first()
          .where('routineId', initRoutines[0].routineId);

        expect(routine).toMatchSnapshot({
          createdAt: expect.any(Date),
          updatedAt: expect.any(Date),
        });

        spy.mockRestore();
      });
    });

    describe('DELETE', () => {
      it('should return 404 if routineId is unknow', async () => {
        const response = await request(app)
          .delete('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2')
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(404);
        expect(response.body).toEqual({});
      });

      it('should return 401 when user is not authenticated', async () => {
        const response = await request(app)
          .delete(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', 'fake');

        expect(response.status).toBe(401);
        expect(response.body).toEqual({});
      });

      it('should delete and stop a routine', async () => {
        const spy = jest.spyOn(RoutineService, 'stop');

        const response = await request(app)
          .delete(`/api/routines/${initRoutines[0].routineId}`)
          .set('Accept', 'application/json')
          .set('x-access-token', user.token);

        expect(response.status).toBe(204);
        expect(RoutineService.stop).toHaveBeenCalledWith(
          initRoutines[0].routineId,
        );

        const routines = await knex(Routine.TABLE)
          .select()
          .where('routineId', initRoutines[0].routineId);

        expect(routines).toHaveLength(0);

        spy.mockRestore();
      });
    });
  });

  describe('/api/routines/:routineId/action', () => {
    it('should run the routine', async () => {
      const spy = jest.spyOn(Routine, 'run');
      const response = await request(app)
        .get(`/api/routines/${initRoutines[0].routineId}/action`)
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(200);
      expect(response.body).toEqual({});
      expect(Routine.run).toHaveBeenCalled();
      expect(Routine.run.mock.calls).toMatchSnapshot();

      spy.mockRestore();
    });

    it('should return 404 if routineId is unknow', async () => {
      const response = await request(app)
        .get('/api/routines/faaed78e-fd1c-4717-b610-65d2fa3d01b2/action')
        .set('Accept', 'application/json')
        .set('x-access-token', user.token);

      expect(response.status).toBe(404);
      expect(response.body).toEqual({});
    });

    it('should return 401 when user is not authenticated', async () => {
      const response = await request(app)
        .get(`/api/routines/${initRoutines[0].routineId}/action`)
        .set('Accept', 'application/json')
        .set('x-access-token', 'fake');

      expect(response.status).toBe(401);
      expect(response.body).toEqual({});
    });
  });
});
