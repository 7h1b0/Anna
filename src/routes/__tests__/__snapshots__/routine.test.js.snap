// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routine API /api/routines GET should retun all routines 1`] = `
Array [
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": true,
    "failReason": null,
    "interval": "0 5 * * *",
    "lastFailedAt": 0,
    "lastRunAt": 0,
    "name": "test at 5am",
    "nextRunAt": 0,
    "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "runAtBankHoliday": true,
    "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
  Object {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": true,
    "failReason": null,
    "interval": "0 9 * * *",
    "lastFailedAt": 0,
    "lastRunAt": 0,
    "name": "test at 9am",
    "nextRunAt": 0,
    "routineId": "1fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "runAtBankHoliday": false,
    "sceneId": "10c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
]
`;

exports[`Routine API /api/routines POST should save and start a new routine 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "failReason": null,
  "interval": "* * * * * *",
  "lastFailedAt": 0,
  "lastRunAt": 0,
  "name": "new_routine",
  "nextRunAt": Any<Number>,
  "routineId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "runAtBankHoliday": true,
  "sceneId": "faaed78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": Any<Number>,
}
`;

exports[`Routine API /api/routines/:routineId GET should return one routine 1`] = `
Object {
  "createdAt": 1514764800000,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "failReason": null,
  "interval": "0 5 * * *",
  "lastFailedAt": 0,
  "lastRunAt": 0,
  "name": "test at 5am",
  "nextRunAt": 0,
  "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "runAtBankHoliday": true,
  "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": 1514851200000,
}
`;

exports[`Routine API /api/routines/:routineId PATCH should update a routine 1`] = `
Array [
  Array [
    Object {
      "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
      "enabled": true,
      "failReason": null,
      "interval": "* * * * * *",
      "lastFailedAt": 0,
      "lastRunAt": 0,
      "name": "routine_updated",
      "nextRunAt": 0,
      "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
      "runAtBankHoliday": true,
      "sceneId": "faaed78e-fd1c-4717-b610-65d2fa3d01b2",
    },
  ],
]
`;

exports[`Routine API /api/routines/:routineId PATCH should update a routine 2`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "failReason": null,
  "interval": "* * * * * *",
  "lastFailedAt": 0,
  "lastRunAt": 0,
  "name": "routine_updated",
  "nextRunAt": 0,
  "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "runAtBankHoliday": true,
  "sceneId": "faaed78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": Any<Number>,
}
`;

exports[`Routine API /api/routines/:routineId/action should run the routine 1`] = `
Array [
  Array [
    RowDataPacket {
      "createdAt": 1514764800000,
      "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
      "enabled": true,
      "failReason": null,
      "interval": "0 5 * * *",
      "lastFailedAt": 0,
      "lastRunAt": 0,
      "name": "test at 5am",
      "nextRunAt": 0,
      "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
      "runAtBankHoliday": true,
      "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
      "updatedAt": 1514851200000,
    },
  ],
]
`;
